tests/tools/oecertdump/enc/enc.cpp

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/entropy.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/rsa.h>
+#include <openenclave/edger8r/enclave.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/report.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "oecertdump_t.h"
+
+// This is the identity validation callback. A TLS connecting party (client or
+// server) can verify the passed in identity information to decide whether to
+// accept a connection request
+oe_result_t enclave_identity_verifier(oe_identity_t* identity, void* arg)
+{
+    oe_result_t result = OE_VERIFY_FAILED;
+
+    OE_UNUSED(arg);
+    OE_TRACE_INFO("enclave_identity_verifier is called with parsed report:\n");
+
+    // Check the enclave's security version
+    if (identity->security_version < 1)
+    {
+        OE_TRACE_ERROR(
+            "identity->security_version checking failed (%d)\n",
+            identity->security_version);
+        goto done;
+    }
+
+    // Dump an enclave's unique ID, signer ID and Product ID. They are
+    // MRENCLAVE, MRSIGNER and ISVPRODID for SGX enclaves. In a real scenario,
+    // custom id checking should be done here
+
+    OE_TRACE_INFO("identity->signer_id :\n");
+    for (int i = 0; i < OE_UNIQUE_ID_SIZE; i++)
+        OE_TRACE_INFO("0x%0x ", (uint8_t)identity->signer_id[i]);
+
+    OE_TRACE_INFO("\nparsed_report->identity.signer_id :\n");
+    for (int i = 0; i < OE_SIGNER_ID_SIZE; i++)
+        OE_TRACE_INFO("0x%0x ", (uint8_t)identity->signer_id[i]);
+
+    OE_TRACE_INFO("\nidentity->product_id :\n");
+    for (int i = 0; i < OE_PRODUCT_ID_SIZE; i++)
+        OE_TRACE_INFO("0x%0x ", (uint8_t)identity->product_id[i]);
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+// input: input_data and input_data_len
+// output: key, key_size
+oe_result_t generate_key_pair(
+    int key_type,
+    uint8_t** public_key,
+    size_t* public_key_size,
+    uint8_t** private_key,
+    size_t* private_key_size)
+{
+    oe_result_t result = OE_FAILURE;
+    oe_asymmetric_key_params_t params;
+    char user_data[] = "test user data!";
+    size_t user_data_size = sizeof(user_data) - 1;
+    uint8_t* local_public_key = NULL;
```

Comment:
nit: .cpp files should normally use the C++ keyword "nullptr" instead of NULL

Reply:
I’m surprised clang-format doesn’t catch this...